File: room/integration-tests/testapp/src/androidTest/java/android/arch/persistence/room/integration/testapp/test/TestUtil.java
/*
 * Copyright (C) 2016 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package android.arch.persistence.room.integration.testapp.test;

import android.arch.persistence.room.integration.testapp.vo.Address;
import android.arch.persistence.room.integration.testapp.vo.Coordinates;
import android.arch.persistence.room.integration.testapp.vo.Pet;
import android.arch.persistence.room.integration.testapp.vo.School;
import android.arch.persistence.room.integration.testapp.vo.User;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.UUID;

public class TestUtil {
    public static User[] createUsersArray(int... ids) {
        User[] result = new User[ids.length];
        for (int i = 0; i < ids.length; i++) {
            result[i] = createUser(ids[i]);
        }
        return result;
    }

    public static List<User> createUsersList(int... ids) {
        List<User> result = new ArrayList<>();
        for (int id : ids) {
            result.add(createUser(id));
        }
        return result;
    }

    public static User createUser(int id) {
        User user = new User();
        user.setId(id);
        user.setName(UUID.randomUUID().toString());
        user.setLastName(UUID.randomUUID().toString());
        user.setAge((int) (10 + Math.random() * 50));
        user.setCustomField(UUID.randomUUID().toString());
        user.setBirthday(new Date());
        return user;
    }

    public static Pet createPet(int id) {
        Pet pet = new Pet();
        pet.setPetId(id);
        pet.setName(UUID.randomUUID().toString());
        return pet;
    }

    public static Pet[] createPetsForUser(int uid, int petStartId, int count) {
        Pet[] pets = new Pet[count];
        for (int i = 0; i < count; i++) {
            Pet pet = createPet(petStartId++);
            pet.setUserId(uid);
            pets[i] = pet;
        }
        return pets;
    }

    public static School createSchool(int id, int managerId) {
        School school = new School();
        school.setId(id);
        school.setName(UUID.randomUUID().toString());
        school.setManager(createUser(managerId));
        school.setAddress(createAddress());
        return school;
    }

    private static Address createAddress() {
        Address address = new Address();
        address.setCoordinates(createCoordinates());
        address.setPostCode((int) (Math.random() * 1000 + 1000));
        address.setState(UUID.randomUUID().toString().substring(0, 2));
        address.setStreet(UUID.randomUUID().toString());
        return address;
    }

    private static Coordinates createCoordinates() {
        Coordinates coordinates = new Coordinates();
        coordinates.lat = Math.random();
        coordinates.lng = Math.random();
        return coordinates;
    }
}

File: appfunctions/integration-tests/multi-modules-testapp/app/src/androidTest/java/androidx/appfunctions/integration/tests/TestUtil.kt
/*
 * Copyright 2025 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.appfunctions.integration.tests

import android.content.ContentValues
import android.content.Context
import android.net.Uri
import android.os.Bundle
import com.google.common.truth.Truth.assertThat
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.runBlocking
import org.junit.Assert.fail

internal object TestUtil {
    fun doBlocking(block: suspend CoroutineScope.() -> Unit) = runBlocking(block = block)

    fun interface ThrowRunnable {
        @Throws(Throwable::class) suspend fun run()
    }

    /** Retries an assertion with a delay between attempts. */
    @Throws(Throwable::class)
    suspend fun retryAssert(runnable: ThrowRunnable) {
        var lastError: Throwable? = null

        for (attempt in 0 until RETRY_MAX_INTERVALS) {
            try {
                runnable.run()
                return
            } catch (e: Throwable) {
                lastError = e
                delay(RETRY_CHECK_INTERVAL_MILLIS)
            }
        }
        throw lastError!!
    }

    fun Context.assertPersistedGranted(uri: Uri) {
        val contentResolver = getContentResolver()
        val targetPersistedUri =
            contentResolver.persistedUriPermissions.singleOrNull { uriPermission ->
                uriPermission.uri == uri
            }
        assertThat(targetPersistedUri).isNotNull()
    }

    fun Context.assertNotPersistedGranted(uri: Uri) {
        val contentResolver = getContentResolver()
        val targetPersistedUri =
            contentResolver.persistedUriPermissions.singleOrNull { uriPermission ->
                uriPermission.uri == uri
            }
        assertThat(targetPersistedUri).isNull()
    }

    /** Asserts that the [Context] having read access to [uri]. */
    fun Context.assertReadAccessible(uri: Uri) {
        val contentResolver = getContentResolver()
        try {
            contentResolver.openAssetFile(uri, "r", null).use { fd ->
                if (fd != null) {
                    return
                }
            }
        } catch (_: Exception) {}
        fail("Uri $uri is not read accessible from $packageName")
    }

    /** Asserts that the [Context] not having read access to [uri]. */
    fun Context.assertReadInaccessible(uri: Uri) {
        val contentResolver = getContentResolver()
        try {
            contentResolver.openAssetFile(uri, "r", null).use { fd -> }
        } catch (_: SecurityException) {
            return
        }
        fail("Uri $uri is still read accessible from $packageName")
    }

    /** Asserts that the [Context] having write access to [uri]. */
    fun Context.assertWriteAccessible(uri: Uri) {
        val contentResolver = getContentResolver()
        try {
            val result =
                contentResolver.update(
                    uri,
                    ContentValues().apply { put("echo_value", 100) },
                    Bundle.EMPTY,
                )
            if (result == 100) {
                return
            }
        } catch (_: Exception) {}
        fail("Uri $uri is not write accessible from $packageName")
    }

    /** Asserts that the [Context] not having write access to [uri]. */
    fun Context.assertWriteInaccessible(uri: Uri) {
        val contentResolver = getContentResolver()
        try {
            contentResolver.update(
                uri,
                ContentValues().apply { put("echo_value", 100) },
                Bundle.EMPTY,
            )
        } catch (_: Exception) {
            return
        }
        fail("Uri $uri is still write accessible from $packageName")
    }

    private const val RETRY_CHECK_INTERVAL_MILLIS: Long = 500
    private const val RETRY_MAX_INTERVALS: Long = 10
}
